# Excercises

### Problem 2
><br>a) <p>S{e}: 8/10 = 0.8
<br>S{b,d}: 2/10 = 0.2
<br> S{b,d,e}: 2/10 = 0.2</p>

><br>b) <p>C {b,d -> e}: 0.2/0.2 = 1
<br> C {e -> b,d}: 0.2/0.8 = 0.25
<br> The above shows that confidence is NOT symmetric.</p>

><br>c) <p>S{e}: 4/5 = 0.8
<br>S{b,d}: 5/5 = 1
<br>S{b,d,e}: 4/5 = 0.8</p>

><br>d) <p>C{b,d -> e}: 0.8/1 = 0.8
<br> C{e -> b,d}: 0.8/0.8 = 1</p>

><br>e) 
# Answer this later

### Problem 6
><br>a)<p>R = 3<sup>6</sup> - 2<sup>7</sup> + 1 = 602</p>

><br>b) <p>Minsup is > 0; therefore, all the itemsets present are frequent. Becuase of transactions 6 and 9 we have a maximum itemset size of 4.p>

><br>c) <p>6 distinct items so there are a possible 6 choose 3 which equals 20.</p>

><br>d) 
<table>
  <tr>
    <th>Itemset</th>
    <th>Support Count</th>
  </tr><tr>
    <td>{Milk, Beer}</td>
    <td>1</td>
  </tr><tr>
    <td>{Milk, Bread}</td>
    <td>3</td>
  </tr><tr>
    <td>{Milk, Butter}</td>
    <td>2</td>
  </tr><tr>
    <td>{Milk, Cookies}</td>
    <td>1</td>
  </tr><tr>
    <td>{Milk, Diapers}</td>
    <td>4</td>
    </tr>
  <tr>
    <td>{Beer, Bread}</td>
    <td>0</td>
  </tr><tr>
    <td>{Beer, Butter}</td>
    <td>0</td>
  </tr><tr>
    <td>{Beer, Cookies}</td>
    <td>2</td>
  </tr><tr>
    <td>{Beer, Diapers}</td>
    <td>3</td>
  </tr><tr>
    <td>{Bread, Butter}</td>
    <td>5</td>
  </tr><tr>
    <td>{Bread, Cookies}</td>
    <td>1</td>
  </tr><tr>
    <td>{Bread, Diapers}</td>
    <td>2</td>
  </tr><tr>
    <td>{Diapers, Butter}</td>
    <td>3</td>
  </tr><tr>
    <td>{Diapers, Cookies}</td>
    <td>1</td>
  </tr><tr>
    <td>{Butter, Cookies}</td>
    <td>1</td>
  </tr>
</table>
<br> {Bread, Butter} is the highest 2 itemset with support 5 and no 3 itemsets have a support greater than 5, and by Apriori's rule no 4 itemsets can have a support greater than 5, so {Bread, Butter} has the largest support of itemsets size >=2.

><br>e) <p>C{Bread -> Butter}: S{bread,butter}/S{bread} = 5/5 = 1
<br> C{Butter -> Bread}: S{Bread, Butter}/S{Butter} = 5/5 = 1
<br> {Bread, Butter} is a pair with the same C{a -> b} and C{b -> a}</p>


### Problem 8
><br>a) <p>{1,2,3,4}, {1,2,3,5}, {1,2,3,6}, {1 2,4,5}, {1,2,4,6}, {1,2,5,6}, {1,3,4,5}, {1,3,4,6}, {2,3,4,5}, {2,3,4,6}, {2,3,5,6}</p>

><br>b)<p>{1,2,3,4}, {1 2,3,5}, {1,2,4,5}, {2,3,4,5}, {2,3,4,6}</p>

><br>c)<p>{1,2,3,4}</p>

### Problem 9
><br>a) 

![image](./diagrams/HW3_Prob9.png)

><br>b)<p>There are 16 frequent itemsets including the null node out of a total of 32,therefore 16/32 or 50% are frequent.</p>

><br>c)<p>There were 11 itemsets pruned therefore the pruning ratio is 11/32 or 34%.</p>

><br>d)<p>There were 5 itemsets that were found to be infrequent during support counting so ther is a false alarm rate of 5/32 or 16%.</p>


### Problem 12

![image](./diagrams/HW3_Prob13.png)

### Problem 13
><br>a) Contingency Tables:

<br> {b}->{c}
<table>
  <tr>
    <th></th>
    <th>c</th>
    <th>c'</th>
  </tr><tr>
    <th>b</td>
    <td>3</td>
    <td>4</td>
  </tr><tr>
    <th>b'</td>
    <td>2</td>
    <td>1</td>
  </tr>
</table>

<br> {a}->{d}
<table>
  <tr>
    <th></th>
    <th>d</th>
    <th>d'</th>
  </tr><tr>
    <th>a</td>
    <td>4</td>
    <td>1</td>
  </tr><tr>
    <th>a'</td>
    <td>5</td>
    <td>0</td>
  </tr>
</table>

<br> {b}->{d}
<table>
  <tr>
    <th></th>
    <th>d</th>
    <th>d'</th>
  </tr><tr>
    <th>b</td>
    <td>6</td>
    <td>1</td>
  </tr><tr>
    <th>b'</td>
    <td>3</td>
    <td>0</td>
  </tr>
</table>

<br> {e}->{c}
<table>
  <tr>
    <th></th>
    <th>c</th>
    <th>c'</th>
  </tr><tr>
    <th>e</td>
    <td>2</td>
    <td>4</td>
  </tr><tr>
    <th>e'</td>
    <td>3</td>
    <td>1</td>
  </tr>
</table>

<br> {c}->{a}
<table>
  <tr>
    <th></th>
    <th>a</th>
    <th>a'</th>
  </tr><tr>
    <th>c</td>
    <td>2</td>
    <td>3</td>
  </tr><tr>
    <th>c'</td>
    <td>3</td>
    <td>2</td>
  </tr>
</table>

><br>b)

<br>Support
<table>
  <tr>
    <th>Rank</th>
    <th>Rule</th>
    <th>Support</th>
  </tr><tr>
    <td>1</td>
    <td>b->d</td>
    <td>0.6</td>
  </tr><tr>
    <td>2</td>
    <td>a->d</td>
    <td>0.4</td>
  </tr><tr>
    <td>3</td>
    <td>b->c</td>
    <td>0.3</td>
  </tr><tr>
    <td>4</td>
    <td>e->c</td>
    <td>0.2</td>
  </tr><tr>
    <td>4</td>
    <td>c->a</td>
    <td>0.2</td>
  </tr>
</table>

<br>Confidence
<table>
  <tr>
    <th>Rank</th>
    <th>Rule</th>
    <th>Confidence</th>
  </tr><tr>
    <td>1</td>
    <td>b->d</td>
    <td>6/7</td>
  </tr><tr>
    <td>2</td>
    <td>a->d</td>
    <td>4/5</td>
  </tr><tr>
    <td>3</td>
    <td>b->c</td>
    <td>3/7</td>
  </tr><tr>
    <td>4</td>
    <td>c->a</td>
    <td>2/5</td>
  </tr><tr>
    <td>5</td>
    <td>e->c</td>
    <td>2.6</td>
  </tr>
</table>

<br>Interest (X->Y)
<table>
  <tr>
    <th>Rank</th>
    <th>Rule</th>
    <th>Interest</th>
  </tr><tr>
    <td>1</td>
    <td>b->d</td>
    <td>0.77</td>
  </tr><tr>
    <td>2</td>
    <td>a->d</td>
    <td>0.72</td>
  </tr><tr>
    <td>3</td>
    <td>b->c</td>
    <td>0.21</td>
  </tr><tr>
    <td>4</td>
    <td>c->a</td>
    <td>0.2</td>
  </tr><tr>
    <td>5</td>
    <td>e->c</td>
    <td>0.17</td>
  </tr>
</table>

<br>IS(X->Y)
<table>
  <tr>
    <th>Rank</th>
    <th>Rule</th>
    <th>IS</th>
  </tr><tr>
    <td>1</td>
    <td>b->d</td>
    <td>0.76</td>
  </tr><tr>
    <td>2</td>
    <td>a->d</td>
    <td>0.60</td>
  </tr><tr>
    <td>3</td>
    <td>b->c</td>
    <td>0.51</td>
  </tr><tr>
    <td>4</td>
    <td>c->a</td>
    <td>0.4</td>
  </tr><tr>
    <td>5</td>
    <td>e->c</td>
    <td>0.37</td>
  </tr>
</table>

<br>Klosgen(X->Y)
<table>
  <tr>
    <th>Rank</th>
    <th>Rule</th>
    <th>Klosgen</th>
  </tr><tr>
    <td>1</td>
    <td>b->d</td>
    <td>-0.033</td>
  </tr><tr>
    <td>2</td>
    <td>b->c</td>
    <td>-0.039</td>
  </tr><tr>
    <td>3</td>
    <td>c->a</td>
    <td>-0.045</td>
  </tr><tr>
    <td>4</td>
    <td>a->d</td>
    <td>-0.063</td>
  </tr><tr>
    <td>5</td>
    <td>e->c</td>
    <td>-0.075</td>
  </tr>
</table>

<br>Odds Ratio
<table>
  <tr>
    <th>Rank</th>
    <th>Rule</th>
    <th>Odds Ratio</th>
  </tr><tr>
    <td>1</td>
    <td>c->a</td>
    <td>0.444</td>
  </tr><tr>
    <td>2</td>
    <td>b->c</td>
    <td>0.375</td>
  </tr><tr>
    <td>3</td>
    <td>e->c</td>
    <td>0.167</td>
  </tr><tr>
    <td>4</td>
    <td>a->d</td>
    <td>0</td>
  </tr><tr>
    <td>4</td>
    <td>b->d</td>
    <td>0</td>
  </tr>
</table>

### Problem 20
><br>a)
<br>Support: S{A} = 0.1, S{B} = 0.9, S{A,B} = 0.09
<br>Interest: 0.09 / (0.9 \* 0.1) = 1
<br>Correlation Coefficient: (f<sub>11</sub>f<sub>00</sub> - f<sub>01</sub>f<sub>10</sub>) / sqrt(f<sub>1+</sub>f<sub>+1</sub>f<sub>0+</sub>f<sub>+0</sub>) = 0.89
<br>Confidence: C{A->B} = 9/10 = 0.9, C{B->A} = 9/10 = 0.9

><br>b)
<br>Support: S{A} = 0.9, S{B} = 0.9, S{A,B} = 0.89
<br>Interest: 1.1
<br>Correlation Coefficient: 0.89
<br>Confidence: C{A->B} = 0.98, C{B->A} = 0.98

><br>c)<p>Support, Confidence, and interest are non-invariant. The Correleation Coefficient is invariant because it checks for the absence of an item in addition to its presence in a transaction.</p>

# Practicum

In [6]:
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

df = pd.read_excel('http://archive.ics.uci.edu/ml/machine-learning-databases/00352/Online%20Retail.xlsx')
df.head()

df['Description'] = df['Description'].str.strip()
df.dropna(axis=0, subset=['InvoiceNo'], inplace=True)
df['InvoiceNo'] = df['InvoiceNo'].astype('str')
df = df[~df['InvoiceNo'].str.contains('C')]

basket = (df[df['Country'] =="France"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))

In [49]:
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1

basket_sets = basket.applymap(encode_units)
basket_sets.drop('POSTAGE', inplace=True, axis=1)

frequent_itemsets = apriori(basket_sets, min_support=0.05, use_colnames=True)
frequent_itemsets.sort_values(by=["support"], ascending=False)


support                                           itemsets
46  0.188776                               (RABBIT NIGHT LIGHT)
52  0.181122                    (RED TOADSTOOL LED NIGHT LIGHT)
44  0.170918                 (PLASTERS IN TIN WOODLAND ANIMALS)
40  0.168367                    (PLASTERS IN TIN CIRCUS PARADE)
59  0.158163               (ROUND SNACK BOXES SET OF4 WOODLAND)
..       ...                                                ...
6   0.051020                            (BLUE HARMONICA IN BOX)
32  0.051020                   (MINI LIGHTS WOODLAND MUSHROOMS)
34  0.051020                    (PACK OF 20 NAPKINS RED APPLES)
70  0.051020                          (SPACEBOY CHILDRENS BOWL)
79  0.051020  (ROUND SNACK BOXES SET OF4 WOODLAND, ALARM CLO...

[113 rows x 2 columns]

The itemset with the greatest support is {Rabbit Night Light} with a support of 0.188776

In [59]:
rules1 = association_rules(frequent_itemsets, metric="confidence", min_threshold=.7)
rules1.sort_values(by=["confidence"], ascending=False)


antecedents  \
24  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...   
25  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...   
17                    (SET/6 RED SPOTTY PAPER PLATES)   
7                        (CHILDRENS CUTLERY SPACEBOY)   
10                     (PACK OF 6 SKULL PAPER PLATES)   
6                      (CHILDRENS CUTLERY DOLLY GIRL)   
18                      (SET/6 RED SPOTTY PAPER CUPS)   
20  (ALARM CLOCK BAKELIKE PINK, ALARM CLOCK BAKELI...   
19  (ALARM CLOCK BAKELIKE PINK, ALARM CLOCK BAKELI...   
3                          (ALARM CLOCK BAKELIKE RED)   
2                        (ALARM CLOCK BAKELIKE GREEN)   
23  (SET/6 RED SPOTTY PAPER PLATES, SET/6 RED SPOT...   
21  (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...   
15                    (SET/6 RED SPOTTY PAPER PLATES)   
9                        (PACK OF 6 SKULL PAPER CUPS)   
5                          (ALARM CLOCK BAKELIKE RED)   
26                    (SET/6 RED SPOTTY PAPER PLATES)   
22  (PLASTERS IN TIN CIRCUS PARADE, PLASTERS IN TI...   
16               (SET/20 RED RETROSPOT PAPER NAPKINS)   
14               (SET/20 RED RETROSPOT PAPER NAPKINS)   
1                        (ALARM CLOCK BAKELIKE GREEN)   
12                         (PLASTERS IN TIN SPACEBOY)   
28               (SET/20 RED RETROSPOT PAPER NAPKINS)   
13                      (SET/6 RED SPOTTY PAPER CUPS)   
4                         (ALARM CLOCK BAKELIKE PINK)   
0                         (ALARM CLOCK BAKELIKE PINK)   
27                      (SET/6 RED SPOTTY PAPER CUPS)   
11                        (PLASTERS IN TIN STRONGMAN)   
8                              (DOLLY GIRL LUNCH BOX)   

                                          consequents  antecedent support  \
24                      (SET/6 RED SPOTTY PAPER CUPS)            0.102041   
25                    (SET/6 RED SPOTTY PAPER PLATES)            0.102041   
17                      (SET/6 RED SPOTTY PAPER CUPS)            0.127551   
7                      (CHILDRENS CUTLERY DOLLY GIRL)            0.068878   
10                       (PACK OF 6 SKULL PAPER CUPS)            0.056122   
6                        (CHILDRENS CUTLERY SPACEBOY)            0.071429   
18                    (SET/6 RED SPOTTY PAPER PLATES)            0.137755   
20                       (ALARM CLOCK BAKELIKE GREEN)            0.073980   
19                         (ALARM CLOCK BAKELIKE RED)            0.073980   
3                        (ALARM CLOCK BAKELIKE GREEN)            0.094388   
2                          (ALARM CLOCK BAKELIKE RED)            0.096939   
23               (SET/20 RED RETROSPOT PAPER NAPKINS)            0.122449   
21                        (ALARM CLOCK BAKELIKE PINK)            0.079082   
15               (SET/20 RED RETROSPOT PAPER NAPKINS)            0.127551   
9                      (PACK OF 6 SKULL PAPER PLATES)            0.063776   
5                         (ALARM CLOCK BAKELIKE PINK)            0.094388   
26  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...            0.127551   
22                 (PLASTERS IN TIN WOODLAND ANIMALS)            0.089286   
16                    (SET/6 RED SPOTTY PAPER PLATES)            0.132653   
14                      (SET/6 RED SPOTTY PAPER CUPS)            0.132653   
1                         (ALARM CLOCK BAKELIKE PINK)            0.096939   
12                 (PLASTERS IN TIN WOODLAND ANIMALS)            0.137755   
28  (SET/6 RED SPOTTY PAPER PLATES, SET/6 RED SPOT...            0.132653   
13               (SET/20 RED RETROSPOT PAPER NAPKINS)            0.137755   
4                          (ALARM CLOCK BAKELIKE RED)            0.102041   
0                        (ALARM CLOCK BAKELIKE GREEN)            0.102041   
27  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...            0.137755   
11                    (PLASTERS IN TIN CIRCUS PARADE)            0.081633   
8                                (SPACEBOY LUNCH BOX)            0.099490   

    consequent suppo

Association Rule with highest confidence is: {Set/6 Red Spotty Paper Plates, Set/20 Red Retrospot Paper Napkins} -> {Set/6 Red Spotty Paper Cups} with a confidence of 0.975

The antecedant is: {Set/6 Red Spotty Paper Plates, Set/20 Red Retrospot Paper Napkins}
The consequent is: {Set/6 Red Spotty Paper Cups}

In [61]:
rules2 = association_rules(frequent_itemsets, metric="lift", min_threshold=7)
rules2.sort_values(by=["lift"], ascending=False)

antecedents  \
8                        (PACK OF 6 SKULL PAPER CUPS)   
9                      (PACK OF 6 SKULL PAPER PLATES)   
6                      (CHILDRENS CUTLERY DOLLY GIRL)   
7                        (CHILDRENS CUTLERY SPACEBOY)   
10  (ALARM CLOCK BAKELIKE PINK, ALARM CLOCK BAKELI...   
15                         (ALARM CLOCK BAKELIKE RED)   
11  (ALARM CLOCK BAKELIKE PINK, ALARM CLOCK BAKELI...   
14                       (ALARM CLOCK BAKELIKE GREEN)   
2                        (ALARM CLOCK BAKELIKE GREEN)   
3                          (ALARM CLOCK BAKELIKE RED)   
12  (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...   
13                        (ALARM CLOCK BAKELIKE PINK)   
4                         (ALARM CLOCK BAKELIKE PINK)   
5                          (ALARM CLOCK BAKELIKE RED)   
18                    (SET/6 RED SPOTTY PAPER PLATES)   
17  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...   
0                         (ALARM CLOCK BAKELIKE PINK)   
1                        (ALARM CLOCK BAKELIKE GREEN)   
16  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...   
19                      (SET/6 RED SPOTTY PAPER CUPS)   

                                          consequents  antecedent support  \
8                      (PACK OF 6 SKULL PAPER PLATES)            0.063776   
9                        (PACK OF 6 SKULL PAPER CUPS)            0.056122   
6                        (CHILDRENS CUTLERY SPACEBOY)            0.071429   
7                      (CHILDRENS CUTLERY DOLLY GIRL)            0.068878   
10                         (ALARM CLOCK BAKELIKE RED)            0.073980   
15  (ALARM CLOCK BAKELIKE PINK, ALARM CLOCK BAKELI...            0.094388   
11                       (ALARM CLOCK BAKELIKE GREEN)            0.073980   
14  (ALARM CLOCK BAKELIKE PINK, ALARM CLOCK BAKELI...            0.096939   
2                          (ALARM CLOCK BAKELIKE RED)            0.096939   
3                        (ALARM CLOCK BAKELIKE GREEN)            0.094388   
12                        (ALARM CLOCK BAKELIKE PINK)            0.079082   
13  (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...            0.102041   
4                          (ALARM CLOCK BAKELIKE RED)            0.102041   
5                         (ALARM CLOCK BAKELIKE PINK)            0.094388   
18  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...            0.127551   
17                    (SET/6 RED SPOTTY PAPER PLATES)            0.102041   
0                        (ALARM CLOCK BAKELIKE GREEN)            0.102041   
1                         (ALARM CLOCK BAKELIKE PINK)            0.096939   
16                      (SET/6 RED SPOTTY PAPER CUPS)            0.102041   
19  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...            0.137755   

    consequent support   support  confidence       lift  leverage  conviction  
8             0.056122  0.051020    0.800000  14.254545  0.047441    4.719388  
9             0.063776  0.051020    0.909091  14.254545  0.047441   10.298469  
6             0.068878  0.063776    0.892857  12.962963  0.058856    8.690476  
7             0.071429  0.063776    0.925926  12.962963  0.058856   12.535714  
10            0.094388  0.063776    0.862069   9.133271  0.056793    6.565689  
15            0.073980  0.063776    0.675676   9.133271  0.056793    2.855230  
11            0.096939  0.063776    0.862069   8.892922  0.056604    6.547194  
14            0.073980  0.063776    0.657895   8.892922  0.056604    2.706829  
2             0.094388  0.079082    0.815789   8.642959  0.069932    4.916181  
3             0.096939  0.079082    0.837838   8.642959  0.069932    5.568878  
12            0.102041  0.063776    0.806452   7.903226  0.055706    4.639456  
13            0.079082  0.063776    0.625000   7.903226  0.055706    2.455782  
4             0.094388  0.073980    0.725000   7.681081  0.064348    3.293135  
5             0.102041  0.073980    0.783784   7.681081  0.064348    4.153061  
18            0.102041  0.

Association Rule with the highest lift is: {Pack of 6 Skull Paper Cups} -> {Pack of 6 Skull Paper Plates} with a lift of ~14.255

The antecedant is: {Pack of 6 Skull Paper Cups}
The consequent is: {Pack of 6 Skull Paper Plates}

We see that the Association Rule with greatest confidence is not the same as the association rule with greatest lift. 